tensorboard/plugins/debugger_v2/local_data_provider.py

```diff
@@ -0,0 +1,98 @@
+# Copyright 2019 The TensorFlow Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""An implementation of DataProvider that serves tfdbg v2 data.
+
+This implementation is:
+  1. Based on reading data from a DebugEvent file set on the local filesystem.
+  2. Implements only the relevant methods for the debugger v2 plugin, including
+     - list_runs()
+     - read_blob_sequences()
+     - read_blob()
+"""
+
+from tensorboard.data import provider
+
+
+DEFAULT_DEBUGGER_RUN_NAME = "__default_debugger_run__"
+
+
+class LocalDebuggerV2DataProvider(provider.DataProvider):
+    """A DataProvider implementation for tfdbg v2 data on local filesystem.
+
+    In this implementation, `experiment_id` is assumed to be the path to the
+    logdir that contains the DebugEvent file set.
```

Comment:
In our discussion, I said that I’d be fine with drafting a temporary
data provider to unblock development of this plugin while we figure out
the data loading story. But it’s important that that short-term provider
be a _valid_ implementation of the `DataProvider` interface and that the
client interact with it according to the documented APIs, so that we can
seamlessly replace this by a real data provider, which will have no
special logic for the debugger plugin. So, for instance:

  - The log directory shouldn’t be communicated via the experiment ID,
    because the logdir is given via a command-line flag while the
    experiment ID is given by a URL component. (For this reason, the
    experiment ID also cannot contain slashes, which makes it a poor
    container for a filesystem path.) Instead, it probably makes more
    sense to mirror the `MultiplexerDataProvider` and accept the logdir
    at data provider initialization time, then start a background thread
    to reload under that directory, like the multiplexer does. Then an
    instance of the data provider is always associated with a logdir,
    and `list_runs` can list the runs under that logdir that have debug
    data.

  - Methods need to be independent. Calling `list_runs` must not affect
    the future behavior of (say) `list_blob_sequences`. Methods may
    depend on state that’s being mutated by a background thread, as in
    the case of the multiplexer, but the methods shouldn’t themselves
    modify state.

  - The result of `list_runs` must be a list of `provider.Run` values,
    not a list of strings.

  - How is the “default debugger run name” behavior going to map to a
    real provider? Perhaps that bit’s just not implemented yet, in which
    case it’s fine, but I don’t see the plan forward.

Does this make sense? Ideally, once the normal loading flow is complete,
the only change necessary will be to change the `self._data_provider`
assignment to read from `context._data_provider` instead of constructing
a new instance, and everything should Just Work™.

Reply:
@wchargin Thanks for the thoughtful review!

> The log directory shouldn’t be communicated via the experiment ID,
> because the logdir is given via a command-line flag while the
> experiment ID is given by a URL component. (For this reason, the
> experiment ID also cannot contain slashes, which makes it a poor
> container for a filesystem path.) Instead, it probably makes more
> sense to mirror the MultiplexerDataProvider and accept the logdir
> at data provider initialization time, then start a background thread
> to reload under that directory, like the multiplexer does. Then an
> instance of the data provider is always associated with a logdir,
> and list_runs can list the runs under that logdir that have debug
> data.

I have created a new class `DebuggerV2EventMultiplexer` in the new file `debug_event_multiplexer.py`. It loosely mirroring `EventMultiplexer`. As you suggested, it is constructed by using a `logdir`. `experiment_id` no longer serve as the substitute `logdir`.

> Methods need to be independent. Calling list_runs must not affect
> the future behavior of (say) list_blob_sequences. Methods may
> depend on state that’s being mutated by a background thread, as in
> the case of the multiplexer, but the methods shouldn’t themselves
> modify state.

I've made changes such that `list_runs()` calls is "stateless" and does. The `read_blob_sequences()` a `read_blob()` methods, the implementation of which has not started in this PR, will also follow this principle of independence.

> The result of list_runs must be a list of provider.Run values,
> not a list of strings.

Done.

> How is the “default debugger run name” behavior going to map to a
> real provider? Perhaps that bit’s just not implemented yet, in which
> case it’s fine, but I don’t see the plan forward.

The default debugger run name is a short-term hack necessitated by the fact that each logdir currently cannot consist of >1 DebugEvent file sets (the reader class in tensorflow will error out if >1 sets are found.) Going forward, we'll work on improving this by allowing there to be multiple file sets in the same logdir.  The newly added TODO item reflects / tracks that:

```py
# TODO(cais): When tfdbg2 allows there to be multiple DebugEvent file sets in
# the same logdir, replace this magic string with actual run names.
```